parlai/agents/examples/seq2seq.py

```diff
@@ -0,0 +1,210 @@
+            This particular implementation returns the hidden and cell states from the
+            LSTM.
+        """
+        embedded = self.embeddings(input_tokens)
+        _output, hidden = self.lstm(embedded)
+        return hidden
+
+
+class Decoder(nn.Module):
+    """
+    Basic example decoder, consisting of an embedding layer and a 1-layer LSTM with the
+    specified hidden size. Decoder allows for incremental decoding by ingesting the
+    current incremental state on each forward pass.
+
+    Pay particular note to the ``forward``.
+    """
+
+    def __init__(self, embeddings, hidden_size):
+        """
+        Initialization.
+
+        Arguments here can be used to provide hyperparameters.
+        """
+        super().__init__()
+        _vocab_size, self.esz = embeddings.weight.shape
+        self.embeddings = embeddings
+        self.lstm = nn.LSTM(
+            input_size=self.esz, hidden_size=hidden_size, num_layers=1, batch_first=True
+        )
+
+    def forward(self, input, encoder_state, incr_state=None):
+        """
+        Run forward pass.
+
+        :param input:
+            The currently generated tokens from the decoder.
+        :param encoder_state:
+            The output from the encoder module.
+        :parm incr_state:
+            The previous hidden state of the decoder.
+        """
+        embedded = self.embeddings(input)
+        if incr_state is None:
+            # this is our very first call. We want to seed the LSTM with the
+            # hidden state of the decoder
+            state = encoder_state
+        else:
+            # We've generated some tokens already, so we can reuse the existing
+            # decoder state
+            state = incr_state
+
+        # get the new output and decoder incremental state
+        output, incr_state = self.lstm(embedded, state)
+
+        return output, incr_state
+
+
+class ExampleModel(tga.TorchGeneratorModel):
+    """
+    ExampleModel implements the abstract methods of TorchGeneratorModel to define how to
+    re-order encoder states and decoder incremental states.
+
+    It also instantiates the embedding table, encoder, and decoder, and defines the
+    final output layer.
+    """
+
+    def __init__(self, dictionary, esz=256, hidden_size=1024):
+        super().__init__(
+            padding_idx=dictionary[dictionary.null_token],
+            start_idx=dictionary[dictionary.start_token],
+            end_idx=dictionary[dictionary.end_token],
+            unknown_idx=dictionary[dictionary.unk_token],
+        )
+        self.embeddings = nn.Embedding(len(dictionary), esz)
+        self.encoder = Encoder(self.embeddings, hidden_size)
+        self.decoder = Decoder(self.embeddings, hidden_size)
+
+    def output(self, decoder_output):
+        """
+        Perform the final output -> logits transformation.
+        """
+        return F.linear(decoder_output, self.embeddings.weight)
```

Comment:
dumb q-- how is this returning logits if it's just a linear transformation

Reply:
Oy. Very good question. I just checked, and `TransformerGeneratorModel` uses the same terminology: https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L939 @stephenroller Can we still consider these "logits" if they aren't mathematically the log of the odds of the probability, but they're still mathematically used in the same manner as logits for computing losses?

Reply:
Oh, makes sense - just talked to Stephen, and they *are* mathematically logits, because when the output of this function gets passed to `CrossEntropyLoss`, that loss criterion *performs* the softmax on the output, making them mathematically the log-odds of probabilities